### PR TITLE
Fixing evaluation of FlaxFunction to include sin, cos, etc

### PR DESCRIPTION
**Pull request by mikaem**
## Summary

`get_fn` in `jaxfun/pinns/loss.py` could not evaluate a nonlinear elementwise function (`sin`, `cos`, `exp`, `tanh`, ...) applied to a `FlaxFunction`, e.g. `sin(u)` or the chain-rule term `cos(u) * Derivative(u, x)` produced by `sp.diff`. This change adds that evaluation path and shares the JAXFUNC-primitive detection helpers between the Galerkin nonlinear integrator and the PINN loss module instead of duplicating them.

## Types of Changes

- [x] Bug fix
- [x] Refactor / cleanup
- [x] Build / CI
- [ ] New feature
- [ ] Performance improvement
- [ ] Documentation update
- [ ] Other (describe):

## Related Issues

Fixes # \
Refs #

## Description of Changes

- `jaxfun/pinns/loss.py`: `get_fn` gains a branch for `type(f).__name__ in JAX_FUNCTION_BY_NAME and len(f.args) == 1` that recurses into the wrapped argument and applies the matching `jnp` function elementwise. Also raises `NotImplementedError` with a descriptive message for expressions `get_fn` cannot handle (e.g. multi-arg functions like `atan2(u, x)`, or un-`doit`'d derivatives of a wrapped `FlaxFunction`) instead of falling through silently.
- `jaxfun/galerkin/arguments.py`: extracted `is_jaxfunc_leaf`, `is_derivative_of_jaxfunc_leaf`, `is_jaxfunc_primitive` helpers (previously private duplicates lived in `jaxfun/integrators/nonlinear.py`); both the Galerkin nonlinear compiler and the PINN loss module now use the shared versions.
- `jaxfun/utils/common.py`: moved the `_JAX_FUNCTION_BY_NAME` sympy-name -> `jnp` function lookup table out of `jaxfun/integrators/nonlinear.py` into `jaxfun/utils/common.py` as public `JAX_FUNCTION_BY_NAME`, reused by `loss.py`.
- `jaxfun/pinns/module.py`: `BaseModule.__hash__` now hashes `(type(self), self.name)` instead of `hash(nnx.graphdef(self)) + hash(self.name)` — some modules (e.g. `MLP.act_fun`) carry plain lists as static metadata, and `GraphDef.__hash__` requires all metadata to be hashable, so hashing could previously raise.
- `jaxfun/pinns/mesh.py`: added `TimeMarchingDomainType` for `TimeMarchingMesh.get_points`'s `domain` argument (accepts `"initial-time"`/`"end-time"` in addition to the base `DomainType`); added `cast(Line, ...)` at two call sites accessing `.left`/`.right` on a `submeshes[-1]` typed more generally; removed two now-unreachable `raise NotImplementedError` branches in `Line` (the preceding `if/elif` chain is exhaustive over `SampleMethod`, which `_coerce_sample_method` already validates against).
- `pyproject.toml` / `uv.lock`: bump `jax` >=0.10 -> >=0.11, `flax` >=0.12 -> >=0.12.8, `optax` >=0.2.6 -> >=0.2.8, `ruff` >=0.15.0 -> >=0.16.0 (lockfile also drops `chex`/`orbax-export`/`dataclasses-json`/`marshmallow`/`typing-inspect`/`mypy-extensions`/`toolz` transitive deps no longer pulled in by the bumped versions).
- Test files: removed now-unnecessary `# ty:ignore[unresolved-attribute]` comments on `jax.config.jax_enable_x64` / `jax.config.x64_enabled` accesses (resolved by the jax bump); added `tests/pinns/test_loss.py` coverage for `sin`/`exp`/`tanh` of a `FlaxFunction`, composition with `Mul`, the chain-rule-derivative case, and the two `NotImplementedError` cases.
- `examples/DrivenCavity.py` / `examples/Helmholtz_vpinn.py`: unrelated example tuning — `Re` 10.0 -> 100.0, and the VPINN Helmholtz residual changed from a cubic nonlinearity (`w**3`) back to linear (`w`), matching the file's own comment/name. Flagging since these aren't obviously connected to the `FlaxFunction` fix — worth a one-line note on why, or splitting out if unintentional.

## Screenshots / Logs (if applicable)

N/A

## Breaking Changes

- [ ] Yes (describe below)
- [x] No

`JAX_FUNCTION_BY_NAME` is now public API in `jaxfun.utils` (was a private module-level dict in `jaxfun.integrators.nonlinear`); no other public signatures changed.

## Checklist

- [x] Tests added or updated
- [ ] Documentation updated (README, examples, docstrings)
- [x] Added type hints
- [x] Linting & formatting pass locally (`ruff check` on touched files — full `pre-commit run --all-files` not run)
- [x] All new and existing tests pass (`uv run pytest`: 790 passed, 9 skipped; `uv run pytest -m pinn tests/pinns/test_loss.py`: 68 passed)
- [ ] Coverage does not decrease (not measured)
- [x] Git history is clean (single commit)

## Additional Notes

- `ty check` passes on all touched source files.

### CODE REVIEW

File: src/jaxfun/integrators/nonlinear.py [hunk 1]
```diff
@@ -285,29 +268,13 @@ def replace_trial_with_jaxfunction(
     return cast(sp.Expr, expr.replace(trial, jax_func))
 
 
-def _is_jaxfunction_leaf(expr: sp.Basic) -> bool:
-    """Return True when ``expr`` is the base JAXFunction field."""
-    return get_arg(expr) is ArgumentTag.JAXFUNC
-
-
 def _contains_jaxfunction(expr: sp.Basic) -> bool:
```
Comment (augustfe) [inline src/jaxfun/integrators/nonlinear.py]: Remove underscore for consistency?

File: src/jaxfun/pinns/loss.py [hunk 2]
```diff
@@ -1099,6 +1101,30 @@ def res_fun(
...
+            return jnp_fn(inner_fn(x, mod, Js, x_id))
+
+        return res_fun
+
+    if not is_jaxfunc_primitive(f):
+        raise NotImplementedError(
+            f"get_fn cannot evaluate {f!r}: expected a bare FlaxFunction application, "
+            "a Derivative of one, a Mul, a Pow, or one of the elementwise functions in "
+            "JAX_FUNCTION_BY_NAME applied to a single FlaxFunction-containing argument"
+        )
```
Comment (augustfe) [inline src/jaxfun/pinns/loss.py:1126]: Very nice

**Review by augustfe — COMMENTED** [review]
Nice job unifying:)
